delocate/libsana.py

```diff
@@ -48,73 +225,104 @@ def tree_libs(start_path, filt_func=None):
 
     * https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/dyld.1.html  # noqa: E501
     * http://matthew-brett.github.io/pydagogue/mac_runtime_link.html
+
+    .. deprecated:: 0.8
+        This function does not support `@loader_path` and only returns the
+        direct dependencies of the libraries in `start_path`.
     """
-    lib_dict = {}
-    env_var_paths = get_environment_variable_paths()
+    warnings.warn(
+        "tree_libs doesn't support @loader_path and has been deprecated.",
+        DeprecationWarning,
+        stacklevel=2,
+    )
+    lib_dict = {}  # type: Dict[Text, Dict[Text, Text]]
     for dirpath, dirnames, basenames in os.walk(start_path):
         for base in basenames:
-            depending_libpath = realpath(pjoin(dirpath, base))
-            if filt_func is not None and not filt_func(depending_libpath):
+            depending_path = realpath(pjoin(dirpath, base))
+            if filt_func and not filt_func(depending_path):
+                logger.debug("Ignoring dependencies of: %s", depending_path)
                 continue
-            rpaths = get_rpaths(depending_libpath)
-            search_paths = rpaths + env_var_paths
-            for install_name in get_install_names(depending_libpath):
-                # If the library starts with '@rpath' we'll try and resolve it
-                # We'll do nothing to other '@'-paths
-                # Otherwise we'll search for the library using env variables
-                if install_name.startswith('@rpath'):
-                    lib_path = resolve_rpath(install_name, search_paths)
-                elif install_name.startswith('@'):
-                    lib_path = install_name
-                else:
-                    lib_path = search_environment_for_lib(install_name)
-                if lib_path in lib_dict:
-                    lib_dict[lib_path][depending_libpath] = install_name
-                else:
-                    lib_dict[lib_path] = {depending_libpath: install_name}
+            for dependency_path, install_name in get_dependencies(
+                depending_path
+            ):
+                if install_name.startswith("@loader_path/"):
+                    # Support for `@loader_path` would break existing callers.
+                    logger.debug(
+                        "Excluding %s because it has '@loader_path'.",
+                        install_name
+                    )
+                    continue
+                lib_dict.setdefault(dependency_path, {})
+                lib_dict[dependency_path][depending_path] = install_name
     return lib_dict
 
 
-def resolve_rpath(lib_path, rpaths):
-    """ Return `lib_path` with its `@rpath` resolved
-
-    If the `lib_path` doesn't have `@rpath` then it's returned as is.
+def resolve_rpath(lib_path, rpaths, loader_path=None, executable_path=None):
```

Comment:
Doesn't this need a rename now?   Maybe something like `resolve_dep_paths`?

Reply:
This was an update to the already existing function, and if you took out rpaths then this would just be a simple string replace operation.

Reply:
Right - yes - I'm just suggesting that string replace.  Maybe something like `resolve_special_paths` or similar, and then something like:

```python
def resolve_rpath(*args, **kwargs):
    """
    .. deprecated:: 0.8
        Please use `resolve_special_paths` instead of this function.
    """
    warnings.warn(
        "Please use `resolve_special_paths` instead of this function.",
        DeprecationWarning,
        stacklevel=2,
    )
    return resolve_special_paths(*args, **kwargs)
```

Reply:
I'm not sure what to do about it.  Trying to resolve rpaths without including `@loader_path` is fundamentally broken, the old function did not support loader_path and I shouldn't be making loader_path optional in the updated function.

I'll try making it a 2nd function.